### PR TITLE
Fix buffer alignment for SDIO

### PR DESCRIPTION
**Pull request by ldursw**
### Description

In issue #21326 the SD card transfer was failing when the buffer size is 512 and ec42be3 is applied. That commit may have moved stuff just enough to expose an alignment issue in Binary File Transfer.

https://github.com/MarlinFirmware/Marlin/blob/b9065195f1babf00ee453c39410206af90a525ae/Marlin/src/HAL/STM32F1/sdio.cpp#L149

This check would fail for every write the code would do resulting in a 0-byte file.

From https://github.com/ChuckM/stm32f4-sdio-driver/blob/f44c727cf21235bf2d0420397a44e5b37b29851c/sdio.c#L607
```cpp
    /*
     * Copy buffer to our word aligned buffer. Nominally you
     * can just use the passed in buffer and cast it to a
     * uint32_t * but that can cause issues if it isn't 
     * aligned.
     */
```

### Requirements

STM32 board with SDIO support. Tested on a Creality 4.2.2 board.

### Benefits

BFT now works with DMA transfers.

### Related Issues

Issue #21326 
PR #21122

### CODE REVIEW

**Comment by rhapsodyv** [comment]
Something is very weird here. You said: "In issue #21326 the SD card transfer was failing when the buffer size is 512 _**and  ec42be3 is applied**_"

Doubt: before https://github.com/MarlinFirmware/Marlin/commit/ec42be346d5c0d072feb8a1b63ef2fdbd6dc1e98 the alignment fix wasn't need?! 🤔

**Comment by ldursw** [comment]
As weird as it sounds that is correct. I did a bisect from d5b0662 to 5ffa6f3 and found out that at ec42be3 BFT stopped working when compression is enabled.

It didn't make any sense because the commit doesn't touch anything related to SD card. My guess is that prior to that commit the buffer location just happened to be aligned but after that the compiler decided to store it somewhere else that wasn't aligned because it wasn't specified to be.

When compression is disabled the code path is different and the transfer succeeds even after that commit.

**Comment by X-Ryl669** [comment]
That smells like a `#pragma pack(push, 1)` leaking and missing a `#pragma pack(pop)`

File: Marlin/src/feature/binary_stream.h [hunk 1]
```diff
@@ -39,8 +39,13 @@ inline int bs_read_serial(const serial_index_t index) {
...
 #if ENABLED(BINARY_STREAM_COMPRESSION)
   static heatshrink_decoder hsd;
+#ifdef BOTH(ARDUINO_ARCH_STM32F1, SDIO_SUPPORT)
+  // STM32 requires a word-aligned buffer for SD card transfers via DMA
+  static __attribute__((aligned(sizeof(size_t)))) uint8_t decode_buffer[512] = {};
+#else
   static uint8_t decode_buffer[512] = {};
-#endif
+#endif // ARDUINO_ARCH_STM32F1
+#endif // BOTH(ARDUINO_ARCH_STM32F1, SDIO_SUPPORT)
```
Comment (tpruvot) [inline Marlin/src/feature/binary_stream.h]: comment is wrong, its BINARY_STREAM_COMPRESSION

**Comment by ldursw** [comment]
> That smells like a `#pragma pack(push, 1)` leaking and missing a `#pragma pack(pop)`

@X-Ryl669 `decode_buffer` is not in a struct. Does `#pragma push` affect declarations like these too? I did a quick search and couldn't find any unmatched pair.

**Pull request merged**